### ISSUE TITLE
Bug While Syncing Nodes

### ISSUE THREAD

**Issue by lautarodragan** [poetapp/node#63]
I found some works are not downloading correctly from IPFS.

They had `db.storage.attemps >= 5`, meaning they had been tried that many times and all failed. But manually resetting that value in the DB triggered another attempt, which succeeded immediately. 

We'll need to upgrade the retry system to have intervals between retries.

**Comment by kennylavender** [comment]
@lautarodragan

**Comment by kennylavender** [comment]
@lautarodragan Let me know what you think of this solution.

One of the issues with the current retry system is that it's running on an interval and the same file can be in the process of being synced multiple times if the retry count is below 5.

Instead of saving the number of attempts in the database we save a sync status value and then use a function to retry on failure up to 5 times and we have a delay between attempts.

The `syncStatus` value might be
- null
- "not synced"
- "syncing"
- "failed synching"
- "synched"

When the interval runs `downloadNexHash`, it looks for files where `claimId` is null and `syncStatus` is null.

Before attempting to sync, it updates the value to "not synced" so that the interval does not attempt this hash again and cause issues by trying to sync the same file more than once at the same time.

Then we attempt to sync, update `syncStatus` to "syncing", this step may not be necessary but might be nice to have this status for debugging.

Next, if it fails we retry, up to 5 times, each time after a set delay.

If all 5 times fail we update `syncStatus` to "failed synching"

If success we update `syncStatus` to "synced".

I think this should create a more reliable synching process.

**Comment by lautarodragan** [comment]
Sounds good in general. Trying to understand your logic. How/when are `ipfsHash`es that have `status: 'not synced` picked up and tried?

**Comment by kennylavender** [comment]
A single function would be responsible for composing the flow. The function that handles the attempts and retries would be called immediately after the status is set to 'not synced'.

**Comment by lautarodragan** [comment]
Where is that info held? (The info about which ipfsHashes need to be retried)

**Comment by kennylavender** [comment]
The beauty is that the data about which ipfsHashes need to be re-tried does not need to exist.

The retry function itself is responsible for retrying until `n` attempts.

Here is a rxjs example of `retry`
https://www.learnrxjs.io/operators/error_handling/retry.html

importing rxjs at this point would be overkill but there are libraries for promises or we could write on
- [Promise Library Example](https://www.npmjs.com/package/promise-retry)
- [Sagas also have this pattern](https://github.com/redux-saga/redux-saga/blob/master/docs/recipes/README.md#debouncing)

**Comment by kennylavender** [comment]
Whoops, just updated the link to the saga pattern, I accidentally linked to debouncing.

**Comment by lautarodragan** [comment]
What happens if an ipfsHash gets picked, its set to `not synced`, and before it is retried, the process gets killed for whatever reason and I have to restart my node?

**Comment by kennylavender** [comment]
good point, I guess in that case that item would never get synced. Maybe we could update all `not synced` items to null on startup?

**Comment by kennylavender** [comment]
in that case, we should also probably set all the ones marked as `syncing` and "failed synching" to null as well on startup.

**Comment by lautarodragan** [comment]
Makes sense. What happens if we want several nodes running against the same database, say for scalability? @kaosmonk @WesleyCharlesBlake do we need to support that case?

**Comment by lautarodragan** [comment]
After further discussion over Slack, this proposal is good to go. Let's give it a shot and put it to the test!

Speaking of which: we do need a way to test this bug, before and after the solution, to make sure it works. We'll probably need to set up an IPFS server — testing this all locally will probably never fail. For this test, I think we won't need to simulate creating new works. Just downloading all the ones that were already created by Frost will do.  One complexity to keep in mind is that there are works that are permanent failures (no longer available), so even if this solution really fixes the issue, we'll still see a lot of failed downloads.

**Comment by kennylavender** [comment]
After even more discussion over Slack, we have decided to go with @lautarodragan idea.

Here is a high-level description:

Use both two timestamps for each item in the DB, `lastSyncAttempt` and `lastSyncSuccess`.

`STORAGE_SYNC_ATTEMPT_DELAY` will be an environment variable, the default value will be 10 mins in milliseconds.

We attempt to sync items on the current interval system where `(lastSyncAttempt > currentTime + sortageSyncAttemptDelay) && `lastSyncSuccess === null`.

As soon as a node finds an item to attempt and sync it sets the `lastSyncAttempt` timestamp to prevent other nodes and the same node from attempting the same item.
